### ISSUE TITLE
ema, stoch are affected by values outside their area of interest

### ISSUE THREAD

**Issue by xucian** [twopirllc/pandas-ta#535]
**Which version are you running? The lastest version is on Github. Pip is for major releases.**
pandas-ta-0.3.14b0 (main)
pandas-ta-0.3.65b0 (development)

**Do you have _TA Lib_ also installed in your environment?**
TA_Lib-0.4.24-cp38-cp38-win_amd64.whl

**Did you upgrade? Did the upgrade resolve the issue?**
already at the latest version

**Describe the bug**
EMA600's values are influenced by values that are even outside the last 600 observations. Shouldn't each value in an EMA**X** be obtained by the 'last **X** observations'? Why is it that even the ones before the last **X** observations (i.e. that aren't needed for EMA's calculation) affect its values?
The difference isn't much, but if you're working on a 1m timeframe with an ema that has 256k data points (~6 months of data), the last value in that EMA would probably be very different from the last value of an EMA created from just the _most recent_ 600 points (i.e. for simplicity, one with only 1 valid value -- the last one).
Even if the difference wasn't much, my initial question holds. I'd like to disable any internal optimizations so that if EMA600 only needs 600 data points, it shouldn't care about- and be affected by what's _before_ those 600 data points. How should I go about this without reinventing the wheel, i.e. creating a 600-len array for each individual step and create call ta.ema() multiple times?

**To Reproduce**
```python
    ema_len = 600
    n_samples = 601
    samples = np.linspace(40_000, 50_000, n_samples)
    

    def create_ema(_samples):
        from pandas import Series
        import pandas as pd
        import pandas_ta as ta
        ser: Series = ta.ema(pd.Series(_samples, dtype=float), length=ema_len)
        return ser.to_numpy()
    

    def test__pands_ta_pas_outside_ema_window__influences__ema_inside_window():
    
        ema_full = create_ema(samples)
        ema_without_first = create_ema(samples[1:])
    
        assert ema_full[-1] == ema_without_first[-1]
```
Gives:

> Expected :45008.333333333336
> Actual   :45008.33333333333

**Expected behavior**
EMAs values should not be affected by input values other than those mathematically required to calculate it

**Additional context**
If we think about any indicator that requires 'the last x observations', is should be implemented this way:
[Python-like pseudocode]
```python
indicator_values = []
vals_so_far = []
# Example: For EMA200 INDICATOR_MIN_REQUIRED_INPUTS is 200
for i in range(N_INPUTS - INDICATOR_MIN_REQUIRED_INPUTS + 1):
    cur_input = INPUTS[N_INPUTS-i-1]
    vals_so_far.insert(0, cur_input)
    if len(vals_so_far) == INDICATOR_MIN_REQUIRED_INPUTS:
        ind_val = calculate_indicator(vals_so_far)
        indicator_values.insert(0, ind_val)

        # Remove last
        vals_so_far = vals_so_far[:-1]

# If you want to pad with NaNs so that the output is of the same size as the input (this is how EMA already works)
if len(indicator_values) < N_INPUTS:
    indicator_values = [np.nan] * (N_INPUTS - len(indicator_values)) + indicator_values
```
In a nutshell, we move from end to start, and get a mathematically accurate indicator value at each data point.
I assume pandas goes from start to end or uses some arithmetic approximations, and this is happening regardless of whether I pass talib=False or True

**Comment by twopirllc** [comment]
Hello @tfgstudios,

Pandas TA is _largely_ a Python implementation of TA Lib (and some _few_ TradingView indicators) and thus the _default_ mode for this Open Source implementation.

<br>

For the sake of brevity, I am only addressing **ema**.

This bug/feature sounds remarkably similar to Issue #420, TA Lib and it's [Unstable Period](https://ta-lib.org/d_api/ta_setunstableperiod.html) as well as code and documentation of TA Lib's [EMA](https://sourceforge.net/p/ta-lib/code/HEAD/tree/trunk/ta-lib/c/src/ta_func/ta_EMA.c) for more details. I chose to implement TA_MA_CLASSIC computation.

Pandas TA currently has three options for an **ema**, ```help(ta.ema)```. You've have tried two of them (1 & 2).
1. With TA Lib installed, it defaults to TA Lib.
1. If ```talib=False``` or TA Lib is not installed, it yields the same result as 1., as intended and you noted.
1. When ```talib=False``` and ```presma=False``` are arguments. (All three in the code and charts below).
    * This attempts to address:
    * > Expected behavior
EMAs values should not be affected by input values other than those mathematically required to calculate it
1. If none of those suit your purposes, you are _encouraged_ to include your own and submit it (preferably in _numpy/numba_). 😎  There are many implementations of "[**ema**](https://en.wikipedia.org/wiki/Moving_average)" beyond TA Lib's [EMA](https://sourceforge.net/p/ta-lib/code/HEAD/tree/trunk/ta-lib/c/src/ta_func/ta_EMA.c). Unfortunately, I have not had enough time nor support to address them all.

<br>

```python
_df = ta.df.ta.ticker("AA", timed=True)["Close"]

def bad_ema(src=None, length=None, n_samples=601):
    if src is None:
        src = pd.Series(np.linspace(40_000, 50_000, n_samples), dtype=float)

    tal_ema = ta.ema(src, length=length, talib=True)
    pta_presma_ema = ta.ema(src, length=length, talib=False)
    pta_ema = ta.ema(src, length=length, talib=False, presma=False)

    return pd.DataFrame({
        "close": src,
        f"tal_ema{length}": tal_ema,
        f"pta_presma_ema{length}": pta_presma_ema,
        f"pta_ema{length}": pta_ema
    })

def cplot(df, last=None):
    if isinstance(last, int):
        df = df.iloc[-last:,:]
    print(df.shape)
    df.plot(figsize=(16,6), color=["black", "red", "orange", "green"] , grid=True)
```

<br/>

```python
ma_length = 10
n_samples = 50
closedf = _df.iloc[:n_samples].copy()

df = bad_ema(closedf, length=ma_length, n_samples=n_samples)
cplot(df, last=None)
df.tail()
```

<img width="747" alt="Screen Shot 2022-05-24 at 7 32 24 PM" src="https://user-images.githubusercontent.com/18198392/170167217-d41c4488-d196-4bfc-8a42-62ac1572b632.png">

It is clearly evident that _pta_ema10_ (green line), in this case, adheres to "EMAs values should not be affected by input values other than those mathematically required to calculate it" expectation you desire. Which by one definition of **ema** relies on a minimum of two values.

Kind Regards,
KJ

**Comment by tfgstudios** [comment]
Hi @twopirllc 
Thanks for the detailed answer!

I'm not sure I understand this part: '_It is clearly evident that pta_ema10 (green line), in this case, adheres to_'
1. why is clear?
2. why does the green line start from the very beginning while the orange one (and the red one that's supposedly beneath it) starts latter? Any values at positions 0, n-2 should be NaNs for EmaN (since it's not possible to calculate them). This is not important for me, as I always strip the first n-2 values, but it might be a problem performance-wise. I presume it's about your ending statement '_Which by one definition of ema relies on a minimum of two values_', but it doesn't click for me.
3. I tried the presma arg in my code using the latest release and it doesn't look like it'd work, i.e. the last values of 2 emas that start at different positions are different (EmaN where: first starts at i and length N+1, second starts at i+1 and length N. For reference, N is 150 here).
4. Also used unstable_period with 'ALL', and 'EMA' (just to make sure), and I don't get the desired result in any of the trials:
```
unstable_period None, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 0, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 5, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 34, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 64, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 65, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 66, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 100, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 149, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 150, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 151, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 200, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 300, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 600, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 700, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 900, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 1500, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period 2000, 		talib True, 	presma True: 	0 out of 1 are equal
unstable_period None, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 0, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 5, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 34, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 64, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 65, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 66, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 100, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 149, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 150, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 151, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 200, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 300, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 600, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 700, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 900, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 1500, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period 2000, 		talib True, 	presma False: 	0 out of 1 are equal
unstable_period None, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 0, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 5, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 34, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 64, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 65, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 66, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 100, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 149, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 150, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 151, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 200, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 300, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 600, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 700, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 900, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 1500, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period 2000, 		talib False, 	presma True: 	0 out of 1 are equal
unstable_period None, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 0, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 5, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 34, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 64, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 65, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 66, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 100, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 149, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 150, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 151, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 200, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 300, 		talib False, 	presma False: 	0 out of 1 are equal
unstable_period 600, 		talib False, 	presma False: 	0 out of 1 are equal
``` 

Is presma  currently implemented in '0.3.14b0'? I browsed the sources on site-packages locally, but there's no code that's extracting it from kwargs:
```
def ema(close, length=None, talib=None, offset=None, **kwargs):
    """Indicator: Exponential Moving Average (EMA)"""
    # Validate Arguments
    length = int(length) if length and length > 0 else 10
    adjust = kwargs.pop("adjust", False)
    sma = kwargs.pop("sma", True)
    close = verify_series(close, length)
    offset = get_offset(offset)
    mode_tal = bool(talib) if isinstance(talib, bool) else True

    if close is None: return

    # Calculate Result
    if Imports["talib"] and mode_tal:
        from talib import EMA
        ema = EMA(close, length)
    else:
        if sma:
            close = close.copy()
            sma_nth = close[0:length].mean()
            close[:length - 1] = npNaN
            close.iloc[length - 1] = sma_nth
        ema = close.ewm(span=length, adjust=adjust).mean()

    # Offset
    if offset != 0:
        ema = ema.shift(offset)

    # Handle fills
    if "fillna" in kwargs:
        ema.fillna(kwargs["fillna"], inplace=True)
    if "fill_method" in kwargs:
        ema.fillna(method=kwargs["fill_method"], inplace=True)

    # Name & Category
    ema.name = f"EMA_{length}"
    ema.category = "overlap"

    return ema
```

**Comment by twopirllc** [comment]
@tfgstudios,

> I now see that presma is present in the development branch. I'm quite reticent to using that branch. It can have more stability issues, right?

Regarding **ema**, it made more sense to rename the argument ```sma``` to ```presma``` from _v0.3.14_ to _development_.

The _development_ branch is equally stable as it's former self, _v0.3.14_ and better. Whether you decide to use the _development_ branch or not, I will not be supporting _v0.3.14_ as it will get replaced by a future version of the _development_ branch after completing [**TODO**](https://github.com/twopirllc/pandas-ta/tree/development#todo)'s Hilbert Transform Indicators, ```ht_*```, under remaining Indicators.

>  I also remember trying to switch to it in the past, but it returned different dimensioned arrays for some indicators (IIRC, the stoch returned 3 arrays instead of 2, and my code logic relies heavily on it returning 2. I can just ignore the additional array, but makes me wonder if there won't be other subtle but important breaking changes)

This library is more feature rich in comparison to some other TA libraries out there and thus some indicators will have more details/columns included with the result, like **stoch**. It is up to the user to drop or exclude extra columns that has no value to them. Others fork the repo and make adjustments.

> If there's a solution for stoch, I'd prefer it to not alter the current values the stoch produces in a significant way -- I've already trained few hundreds parameters based on it and can't afford retraining the models.

The next time I will be touching it is when I convert it to _numpy/numba_. At the current rate, it won't be anytime soon.

> But my other tests still fail in some cases (described by (*) above).

There are several other TA libraries out there. Have you tried them? I am curious if they have solved "indicator(s) are affected by values outside their area of interest"? 🤔 

KJ

<br/>

### Addendum

> Expected behavior
EMAs values should not be affected by input values other than those mathematically required to calculate it

This is what I hear:
<img width="711" alt="Screen Shot 2022-05-25 at 6 23 31 AM" src="https://user-images.githubusercontent.com/18198392/170272555-5cd82ce0-ec42-45f9-831d-2f0978638734.png">

As shown above, set ```talib=False, presma=False``` only uses two consecutive values as detailed [here](https://en.wikipedia.org/wiki/Moving_average).

This calculation is done by [Panda's **ewm**](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html?highlight=ewm#pandas.DataFrame.ewm).

```python
def ema(*args, **kwargs):
    # ...
    close.ewm(span=length, adjust=adjust).mean()  # where adjust=False
    # ...
```

**Comment by twopirllc** [comment]
Please see #901
